src/compiler/utilitiesPublic.ts

```diff
@@ -304,7 +305,7 @@ export function isExternalModuleNameRelative(moduleName: string): boolean {
 }
 
 export function sortAndDeduplicateDiagnostics<T extends Diagnostic>(diagnostics: readonly T[]): SortedReadonlyArray<T> {
-    return sortAndDeduplicate<T>(diagnostics, compareDiagnostics);
+    return sortAndDeduplicate<T>(diagnostics, compareDiagnostics, diagnosticsEqualityComparer);

```

Comment:
As far as I can tell, `DIagnosticCollection`'s `.add` method is still using `compareDiagnosticsSkipRelatedInformation`, so it's just first-in-wins (with the same root message) - should that also get updated to prefer the "lesser" message like this?

Reply:
`compareDiagnosticsSkipRelatedInformation` is still the same in terms of what it considers equal - two diagnostics with the same location and head message but different elaboration will still be considered different by that function, and so both will be added via `DiagnosticCollection.add`. Only later, when we call `sortAndDeduplicateDiagnostics`, will we use `diagnosticsEqualityComparer` and get rid of all but one of the diagnostics that we now consider equal.

Reply:
Now that I think about it, if we want to keep using functions like `sortAndDeduplicate` and `insertSorted`, I think we need to have two separate functions for comparison and equality: one that fully compares diagnostics for purposes of sorting (and sorting a diagnostic with more elaboration before one with less), and another for purposes of deduplication that is more permissive and only compares location and head message for equality. Otherwise we'd have to implement a function that somehow says "yes, those two diagnostics are equal, but one of them is preferred".